hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/operations/PartitionIteratingOperation.java

```diff
@@ -181,19 +173,52 @@ protected void toString(StringBuilder sb) {
         sb.append(", operationFactory=").append(operationFactory);
     }
 
-    private static class ResponseQueue implements OperationResponseHandler {
+    private class OperationResponseHandlerImpl implements OperationResponseHandler {
 
-        private final BlockingQueue<Object> queue = ResponseQueueFactory.newResponseQueue();
+        // an array with the partitionCount as length, so we can quickly do a lookup for a given partitionId.
+        // it will store all the 'sub' responses.
+        private final AtomicReferenceArray<Object> responseArray = new AtomicReferenceArray<Object>(
+                getNodeEngine().getPartitionService().getPartitionCount());
+
+        // contains the number of pending operations. If it hits zero, all responses have been received.
+        private final AtomicInteger pendingOperations;
+        private final int[] partitions;
+
+        OperationResponseHandlerImpl(int[] partitions) {
+            this.partitions = partitions;
+            this.pendingOperations = new AtomicInteger(partitions.length);
+        }
 
         @Override
-        public void sendResponse(Operation op, Object obj) {
-            if (!queue.offer(obj)) {
-                throw new HazelcastException("Response could not be queued for transportation");
+        public void sendResponse(Operation op, Object response) {
+            if (response instanceof NormalResponse) {
+                response = ((NormalResponse) response).getValue();
+            } else if (response == null) {
+                response = NULL;
+            }
+
+            // we try to set the response in the responseArray
+            if (!responseArray.compareAndSet(op.getPartitionId(), null, response)) {
+                //duplicate response; should not happen.
```

Comment:
What about adding an assertion here? All unit tests and simulator tests are run assertions enabled.